# Import Libraries

In [229]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
import sklearn
from sklearn import tree
from sklearn.preprocessing import Imputer
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import precision_recall_fscore_support, average_precision_score
from sklearn.naive_bayes import GaussianNB
from operator import itemgetter ,attrgetter, methodcaller
from sklearn.svm import LinearSVC
from sklearn  import linear_model
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import RidgeCV
from sklearn.preprocessing import PolynomialFeatures
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import Perceptron
from sklearn.datasets import make_classification
from sklearn.ensemble import ExtraTreesClassifier

Read clean dataset 

In [230]:
df1 = pd.read_csv("E:\ML\Assignment1\Crime Prediction Data\communities-crime-clean.csv")

# 1. Decision Trees

a.	Create a new field “highCrime” which is true if the crime rate per capita (ViolentCrimesPerPop) is greater than 0.1, and false otherwise.

In [231]:
df1["highCrime"] = df1["ViolentCrimesPerPop"] > 0.1

In [232]:
df1

state        communityname  fold  population  householdsize  \
0         1        Alabastercity     7        0.01           0.61   
1         1    AlexanderCitycity    10        0.01           0.41   
2         1         Annistoncity     3        0.03           0.34   
3         1           Athenscity     8        0.01           0.38   
4         1           Auburncity     1        0.04           0.37   
5         1         Bessemercity     6        0.04           0.44   
6         1       Birminghamcity     2        0.41           0.37   
7         1          Cullmancity     1        0.01           0.30   
8         1           Daphnecity     7        0.00           0.39   
9         1          Decaturcity    10        0.06           0.39   
10        1           Dothancity     4        0.07           0.41   
11        1       Enterprisecity     2        0.02           0.43   
12        1          Eufaulacity     5        0.01           0.46   
13        1        Fairfieldcity     1        0.00           0.45   
14        1         Florencecity     3        0.04           0.33   
15        1        FortPaynecity     2        0.00           0.35   
16        1          Gadsdencity     9        0.05           0.32   
17        1        Hartsellecity     4        0.00           0.43   
18        1         Homewoodcity     5        0.02           0.23   
19        1           Hoovercity    10        0.05           0.35   
20        1         Hueytowncity     4        0.01           0.44   
21        1       Huntsvillecity     1        0.24           0.37   
22        1     Jacksonvillecity     7        0.00           0.54   
23        1           Jaspercity     8        0.01           0.37   
24        1          Madisoncity     5        0.01           0.36   
25        1           Mobilecity     8        0.30           0.41   
26        1       Montgomerycity     2        0.28           0.45   
27        1    MountainBrookcity     8        0.02           0.37   
28        1        Northportcity     2        0.01           0.44   
29        1          Opelikacity     6        0.02           0.45   
...     ...                  ...   ...         ...            ...   
1963     55        NewBerlincity     4        0.04           0.55   
1964     55         OakCreekcity     6        0.02           0.49   
1965     55       Oconomowoccity     6        0.00           0.42   
1966     55         Onalaskacity     4        0.00           0.44   
1967     55          Oshkoshcity     7        0.07           0.42   
1968     55           Racinecity    10        0.12           0.44   
1969     55       RiverFallscity     3        0.00           0.67   
1970     55        Sheboygancity     8        0.06           0.37   
1971     55     Shorewoodvillage     6        0.01           0.18   
1972     55   SouthMilwaukeecity     9        0.02           0.39   
1973     55     StevensPointcity     4        0.02           0.51   
1974     55       SunPrairiecity     2        0.01           0.48   
1975     55         Superiorcity     8        0.03           0.34   
1976     55        TwoRiverscity     6        0.00           0.37   
1977     55        Watertowncity     2        0.00           0.52   
1978     55         Waukeshacity     6        0.08           0.45   
1979     55           Wausaucity     7        0.04           0.37   
1980     55        Wauwatosacity     9        0.06           0.35   
1981     55        WestAlliscity     6        0.09           0.29   
1982     55         WestBendcity     3        0.02           0.49   
1983     55  WhitefishBayvillage     9        0.01           0.42   
1984     55       Whitewatercity     6        0.00           0.65   
1985     55  WisconsinRapidscity     8        0.01           0.32   
1986     56         Cheyennecity     4        0.06           0.34   
1987     56         Evanstoncity     6        0.00           0.65   
1988     56         Gillettecity     9        0.01           0.53   
1989

What are the percentage of positive and negative instances in the dataset?

In [233]:
count = df1['highCrime'].value_counts(normalize = True)
count * 100

True     62.719518
False    37.280482
Name: highCrime, dtype: float64

The attribute ‘highCrime’ predicts the  value “true” if true if the crime rate per capita (ViolentCrimesPerPop) is greater than 0.1, and false otherwise.  On calculating the percentage of true and false values, I observed that approximately 62% of given data says that the crime rate per capita is greater than 0.1.


Create the copy of dataset to work on

In [234]:
df2 = df1.copy()

Remove unpredictive columns from dataset

In [235]:
del df2['state']
del df2['communityname']
del df2['fold']
del df2['highCrime']
del df2['ViolentCrimesPerPop']
df2.head()

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.01           0.61          0.21          0.83          0.02   
1        0.01           0.41          0.55          0.57          0.01   
2        0.03           0.34          0.86          0.30          0.04   
3        0.01           0.38          0.35          0.71          0.04   
4        0.04           0.37          0.32          0.70          0.21   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.01         0.41         0.49         0.26        0.21   
1         0.00         0.47         0.45         0.31        0.57   
2         0.01         0.41         0.42         0.27        0.59   
3         0.01         0.39         0.46         0.31        0.49   
4         0.02         1.00         1.00         1.00        0.14   

          ...           NumStreet  PctForeignBorn  PctBornSameState  \
0         ...                0.00            0.03              0.70   
1         ...                0.00            0.00              0.93   
2         ...                0.02            0.04              0.77   
3         ...                0.00            0.03              0.78   
4         ...                0.00            0.12              0.49   

   PctSameHouse85  PctSameCity85  PctSameState85  LandArea  PopDens  \
0            0.40           0.34            0.57      0.05     0.06   
1            0.66           0.82            0.84      0.11     0.03   
2            0.59           0.70            0.64      0.06     0.11   
3            0.56           0.67            0.71      0.09     0.05   
4            0.12           0.00            0.15      0.09     0.09   

   PctUsePubTrans  LemasPctOfficDrugUn  
0            0.01                  0.0  
1            0.01                  0.0  
2            0.04                  0.0  
3            0.00                  0.0  
4            0.01                  0.0  

[5 rows x 100 columns]

b.	 Use DecisionTreeClassifier to learn a decision tree to predict highCrime on the entire dataset 

In [236]:
clf = tree.DecisionTreeClassifier()

i.	What are the training accuracy, precision, and recall for this tree? 

In [237]:
model = DecisionTreeClassifier()
model.fit(df2, df1['highCrime'])
p = model.predict(df2)
acc = model.score(df2, df1['highCrime'])
pre = (precision_recall_fscore_support(df1['highCrime'], p))
print("Accuracy: %f" % acc)
print(pre)


Accuracy: 1.000000
(array([ 1.,  1.]), array([ 1.,  1.]), array([ 1.,  1.]), array([ 743, 1250], dtype=int64))


ii.	What are the main features used for classification? Can you explain why they make sense (or not)?

In [238]:
col = df2.columns
imp_features = model.feature_importances_
indices = np.argsort(imp_features)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print("{}. feature {} ({})".format(i + 1, col[indices[i]], imp_features[indices[i]]))
   

Feature Ranking
1. feature PctKids2Par (0.35865224139473956)
2. feature racePctWhite (0.08646057909879694)
3. feature racePctHisp (0.047066975125015176)
4. feature PctEmplManu (0.02575877159536193)
5. feature PctLess9thGrade (0.020557220758796)
6. feature MedRent (0.015301814701351528)
7. feature PctImmigRec10 (0.014399759238583115)
8. feature PctIlleg (0.014334316344717988)
9. feature HispPerCap (0.012832658084347854)
10. feature PctSameState85 (0.012219169785146995)


The top 10 important features for decision tree classification are observed as follows. The most important feature (PctKids2Par) i.e. percentage of kids with 2 parents hold the maximum weight. The reason for this attribute to be the important feature for highCrime may be that children with 2 parents lack boundaries and parental supervision due to difficult family relationships. This may result in getting the children in company with people who commit crime or misuse drug and alcohol. The attributes racePctWhite and racePctHisp says that these people mostly belong to White or Hispanic race. The attribute PctLess9thGrade indicates the percentage of people with age 25 and over with less than 9th grade education. This attribute is one of the main features of crime rate which indicates that lack of good education is one of the reason for high crime. The attribute PctEmplManu represent people who are employed in manufacturing industry. This manufacturing industry may include drug or weapon manufacturing which easily introduces 
people with these things and may result in misuse of this manufacturing knowledge.

c.	Now apply cross-validation (cross_val_score) to do 10-fold cross-validation to estimate the out-of-training accuracy of decision tree learning for this task.

i.	What are the 10-fold cross-validation accuracy, precision, and recall?


In [239]:
print("Precision: %f" % cross_val_score(model , df2, df1['highCrime'], cv=10, scoring='precision' ).mean())
print ("Recall: %f" %cross_val_score(model , df2, df1['highCrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy: %f" %cross_val_score(model , df2, df1['highCrime'], cv=10, scoring='accuracy' ).mean())

Precision: 0.782561
Recall: 0.775200
Accuracy: 0.731505


The results in this test are different than the earlier test because in earlier test, the model was being trained and tested on the same set of data which resulted in overfitting and giving the 100% result. In 10 fold cross validation method, 10 random parts of data is chosen, out of which 9 are trained and tested on tenth. This is repeated training and testing dataset changes everytime.

# 2. Linear Classification

a.	Use GaussianNB to learn a Naive Bayes classifier to predict highCrime.

i.	What is the 10-fold cross-validation accuracy, precision, and recall for this method?


In [240]:
model1 = GaussianNB()
print("Precision: %f" % cross_val_score(model1, df2, df1['highCrime'], cv=10, scoring='precision' ).mean())
print ("Recall: %f" %cross_val_score(model1, df2, df1['highCrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy: %f" %cross_val_score(model1, df2, df1['highCrime'], cv=10, scoring='accuracy' ).mean())

Precision: 0.911800
Recall: 0.692000
Accuracy: 0.761608


Prediction using Naive Bayes classifier

In [241]:
df3 = df2.copy()
model1.fit(df2, df1['highCrime'])
df3['Prediction'] = model1.predict(df2)
df3

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.01           0.61          0.21          0.83          0.02   
1           0.01           0.41          0.55          0.57          0.01   
2           0.03           0.34          0.86          0.30          0.04   
3           0.01           0.38          0.35          0.71          0.04   
4           0.04           0.37          0.32          0.70          0.21   
5           0.04           0.44          1.00          0.10          0.00   
6           0.41           0.37          1.00          0.02          0.03   
7           0.01           0.30          0.00          0.99          0.02   
8           0.00           0.39          0.31          0.75          0.02   
9           0.06           0.39          0.32          0.73          0.04   
10          0.07           0.41          0.53          0.57          0.05   
11          0.02           0.43          0.41          0.64          0.08   
12          0.01           0.46          0.67          0.47          0.02   
13          0.00           0.45          1.00          0.00          0.01   
14          0.04           0.33          0.33          0.73          0.02   
15          0.00           0.35          0.09          0.91          0.02   
16          0.05           0.32          0.55          0.56          0.04   
17          0.00           0.43          0.12          0.90          0.01   
18          0.02           0.23          0.16          0.85          0.08   
19          0.05           0.35          0.06          0.93          0.07   
20          0.01           0.44          0.26          0.80          0.01   
21          0.24           0.37          0.48          0.58          0.13   
22          0.00           0.54          0.34          0.71          0.07   
23          0.01           0.37          0.28          0.77          0.02   
24          0.01           0.36          0.15          0.83          0.16   
25          0.30           0.41          0.76          0.38          0.06   
26          0.28           0.45          0.83          0.34          0.04   
27          0.02           0.37          0.00          0.99          0.04   
28          0.01           0.44          0.44          0.64          0.05   
29          0.02           0.45          0.76          0.38          0.08   
...          ...            ...           ...           ...           ...   
1963        0.04           0.55          0.00          0.98          0.06   
1964        0.02           0.49          0.01          0.96          0.05   
1965        0.00           0.42          0.00          1.00          0.01   
1966        0.00           0.44          0.00          0.99          0.03   
1967        0.07           0.42          0.02          0.95          0.13   
1968        0.12           0.44          0.36          0.64          0.03   
1969        0.00           0.67          0.01          0.97          0.07   
1970        0.06           0.37          0.00          0.92          0.24   
1971        0.01           0.18          0.03          0.95          0.10   
1972        0.02           0.39          0.01          0.98          0.03   
1973        0.02           0.51          0.01          0.94          0.17   
1974        0.01           0.48          0.01          0.98          0.03   
1975        0.03           0.34          0.01          0.95          0.05   
1976        0.00           0.37          0.00          0.97          0.08   
1977        0.00           0.52          0.00          0.99          0.03   
1978        0.08           0.45          0.01          0.93          0.08   
1979        0.04           0.37          0.00          0.90          0.37   
1980        0.06           0.35          0.02          0.96          0.06   
1981        0.09           0.29          0.01          0.98          0.04   
1982        0.02           0.49          0.00          0.99          0.03   
19

ii. What are the 10 most predictive features? This can be measured by the normalized absolute difference of means for the feature between the two classes:
|μ_T-μ_F |/(σ_T+σ_F )
The larger this different, the more predictive the feature. Why do these make sense (or not)?


In [242]:
dfT = df3[df3['Prediction'] == True]
dfF = df3[df3['Prediction'] == False]
nb = []
for i in range(len(df2.columns)):
    nb.append((abs(dfT[[i]].mean() - dfF[[i]].mean())  / dfT[[i]].std() + dfF[[i]].std()))
sorted(nb, key = itemgetter(0), reverse = True)[:10]

[PctKids2Par    1.881312
 dtype: float64, pctWInvInc    1.841672
 dtype: float64, PctPersOwnOccup    1.769371
 dtype: float64, medFamInc    1.763879
 dtype: float64, PctFam2Par    1.757952
 dtype: float64, medIncome    1.733898
 dtype: float64, FemalePctDiv    1.679538
 dtype: float64, PctYoungKids2Par    1.59169
 dtype: float64, PctHousLess3BR    1.579829
 dtype: float64, TotalPctDiv    1.575829
 dtype: float64]

Here we see that the most predictive feature is based on the above factor involving the means of the two classes used for classification: true, false. The most predictive feature has the highest mean difference. The predictive features are divided on the basis of mean difference. The higher the mean difference the better the classification. The better classification represents that the dataset is divided in purer sets. Thus, using this factor makes sense.

iii.	How do these results compare with your results from decision trees, above?

Decision tree
Precision: 0.793604
Recall: 0.777600
Accuracy: 0.730523

Imp_features:
PctKids2Par (0.3620975950660422)
racePctWhite (0.08646057909879697)


Gaussian Nb 
Precision: 0.911800
Recall: 0.692000
Accuracy: 0.761608

Imp_features:
PctKids2Par    1.881312
pctWInvInc    1.841672

Decision tree and NBGauss, both classify the dataset on the basis of the attribute 'PctKids2Par'. Value of PctKids2Par for decision tree is 0.36 and for NBGauss the value is 1.88.
The accuracy and precision for NBGauss is better than decision tree. This represents that NBGauss classifies the data more true than the decision tree. It is important to know true classification the places with higher crime rate
than just to know that the crime has been predicted. This can be done using recall which is better for decision tree.

b.	Use LinearSVC to learn a linear Support Vector Machine model to predict highCrime.

i.	 What is the 10-fold cross-validation accuracy, precision, and recall for this method?

In [243]:
model2 = LinearSVC()
print("Precision: %f" % cross_val_score(model2, df2, df1['highCrime'], cv=10, scoring='precision' ).mean())
print ("Recall: %f" %cross_val_score(model2, df2, df1['highCrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy: %f" %cross_val_score(model2, df2, df1['highCrime'], cv=10, scoring='accuracy' ).mean())

Precision: 0.845405
Recall: 0.834400
Accuracy: 0.796234


ii.	What are the 10 most predictive features? This can be measured by the absolute feature weights in the model. Why do these make sense (or not)?

In [244]:
model2.fit(df2, df1['highCrime'])
SVC_imp_features = model2.coef_[0]
indices = np.argsort(SVC_imp_features)[::-1]
print('Feature Ranking')
for i in range(10):
    print("{}. feature {} ({})".format(i + 1, col[indices[i]], SVC_imp_features[indices[i]]))

Feature Ranking
1. feature PersPerOccupHous (1.755123507984769)
2. feature RentHighQ (1.0668838199521737)
3. feature MalePctDivorce (1.065696568739363)
4. feature NumUnderPov (1.0515477266798177)
5. feature NumStreet (1.0191543542568045)
6. feature PctOccupMgmtProf (1.0146738514033011)
7. feature population (1.0023003741621772)
8. feature agePct12t21 (0.988272900239368)
9. feature HousVacant (0.9365130844308158)
10. feature agePct65up (0.8905733694826011)


iii.	How do these results compare with your results from decision trees, above?

Support vector machines (SVMs) are a set of supervised learning methods used for classification, regression and outliers detection. It uses a technique called the kernel trick to 
transform your data and then based on these transformations it finds an optimal boundary between the possible outputs. 

Feature selection is a process by which you automatically search for the best subset of attributes in your dataset.  The best features combination gives us highest accuracy which means less misleading of data.


Decision tree
Precision: 0.793604
Recall: 0.777600
Accuracy: 0.730523

imp_features:
PctKids2Par (0.3620975950660422)
racePctWhite (0.08646057909879697)

linearSVC
Precision: 0.845405
Recall: 0.834400
Accuracy: 0.796234

imp_features:
feature PersPerOccupHous (1.7551257948297427)
feature RentHighQ (1.0668846550482343)


Precision, recall and accuracy are better for SVC than for decision tree. SVC hold good threshold and better to predict the features.


# 3. Regression

a.	Use LinearRegression to learn a linear model directly predicting the crime rate per capita (ViolentCrimesPerPop).

i.	Using 10-fold cross-validation, what is the estimated mean-squared-error (MSE) of the model?

In [245]:
model3 = linear_model.LinearRegression()
model3.fit(df2,df1['ViolentCrimesPerPop'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [246]:
predict = model3.predict(df2)
mse_scores = cross_val_score(model3, df2 ,df1['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' , cv=10)
mse_scores.mean()

-0.020093969304445319

ii.	What is the MSE on the training set (train on all the data then test on it all)?

In [247]:
mean_squared_error(df1['ViolentCrimesPerPop'], predict)

0.016516774880307179

When you are testing on the saem data that you trained, the accuracy is maximum. Thus, mean squared error is minimum and gives the best possible result.

iii.	What features are most predictive of a high crime rate? A low crime rate?

In [248]:
important_features = model3.coef_
top_indices = np.argsort(important_features)[::-1]
bottom_indices = np.argsort(important_features)
print ('Top 10 Feature Ranking') 
for i in range(10):
    print ("{}. Feature {} ({})".format(i+1, col[top_indices[i]], important_features[top_indices[i]]))
print('')
print('Bottom 10 Feature Ranking')
for i in range(10):
    print ("{}. Feature {} ({})".format(i+1, col[bottom_indices[i]], important_features[bottom_indices[i]]))


Top 10 Feature Ranking
1. Feature PersPerOccupHous (0.6350881164986149)
2. Feature PctHousOwnOcc (0.5681332098865255)
3. Feature MalePctDivorce (0.45851704864166676)
4. Feature PctRecImmig8 (0.43251055765291013)
5. Feature MedRent (0.3727277977113885)
6. Feature medFamInc (0.2879788749460697)
7. Feature PctEmploy (0.248474316084703)
8. Feature MalePctNevMarr (0.22672791284180477)
9. Feature PctPersDenseHous (0.2143525641419764)
10. Feature OwnOccMedVal (0.21287586365359717)

Bottom 10 Feature Ranking
1. Feature PctPersOwnOccup (-0.6756944788028594)
2. Feature TotalPctDiv (-0.561924314414664)
3. Feature whitePerCap (-0.3510157744407789)
4. Feature PctKids2Par (-0.32265127649627473)
5. Feature OwnOccLowQuart (-0.3081702191931843)
6. Feature numbUrban (-0.29644254317756596)
7. Feature PersPerRentOccHous (-0.2545716627164686)
8. Feature RentLowQ (-0.2347515163025197)
9. Feature agePct12t29 (-0.22921783699885295)
10. Feature PctRecImmig5 (-0.2182210073546513)


Used coefficients of regression models for selecting and interpreting features. This is based on the idea that when all features are on the same scale, the most important features should have the highest coefficients in the model, while features uncorrelated with the output variables should have coefficient values close to zero. This approach can work well even with simple linear regression models when the data is not very noisy.

b.	Now use Ridge regression to reduce the amount of overfitting, using RidgeCV to pick the best alpha from among (10, 1, 0.1, 0.01, and 0.001).

i.	What is the estimated MSE of the model under 10-fold CV?

In [249]:
model4 = linear_model.RidgeCV(alphas = (10, 1, 0.1, 0.01, 0.001))
model4.fit(df2, df1['ViolentCrimesPerPop'])
mse = cross_val_score(model4, df2, df1['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' ,cv=10)
mse.mean()

-0.019795021348168747

ii.	What is the MSE on the training set (train on all the data then test on it all)?

In [250]:
ridge_predict = model4.predict(df2)
mean_squared_error(df1['ViolentCrimesPerPop'], ridge_predict)

0.016763529155169436

iii.	What is the best alpha?

In [251]:
model4.alpha_

1.0

Alpha, also called as regularization, is used to set certain penality to reduse noise over the data. Larger the alpha value, more is the domination effect for regularization which tends the coefficients to zero. If the alpha value tends to be zero and the solution implies towards the ordinary least squares then coefficients tend to be big oscillation. Thus, it is preferable that the alpha is selected in a way that it maintains the balance between both. 

iv.	What does this say about the amount of overfitting in linear regression for this problem?

MSE for Linear Regression: -0.020093969304445319

MSE for Ridge Regression: -0.019795021348168747

Mean Squared Error (MSE) helps to determine if the model is overfitted or underfitted. The MSE value for Linear Regression here is very small thus representing very less overfitting. Linear regression is modeled from the training data coefficients of every feature used.  Thus causing an overfitting problem as each of the declared coefficient estimates will not specify the true patterns and relationship in dataset. It only presents the sample data.
Ridge regression specify prior knowledge about how the regression plays the role in estimation (determined by alpha value)

c.	Now use polynomial features to do quadratic (second-order) polynomial regression.

In [252]:
feature_set = df1[list(df1)].values
labels = df2.values
x_features = np.array(feature_set)
y_labels = np.array(labels)

reg_features = df2[list(df2)].values
reg_labels = df1['ViolentCrimesPerPop'].values
x_reg_features = np.array(reg_features)
y_reg_labels = np.array(reg_labels)

model4 = PolynomialFeatures(degree=2)
X = model4.fit_transform(x_reg_features)


model5 = linear_model.LinearRegression()
model5.fit(X, y_reg_labels)
pred = model4.fit_transform(x_reg_features)
y_predict = model5.predict(pred)

i.	What is the estimated MSE of the model under 10-fold CV?

In [253]:
mse = cross_val_score(model5, x_reg_features,  y_reg_labels, cv=10, scoring='neg_mean_squared_error')
print(" Mean MSE: %f" % cross_val_score(model5 , df2, df1['highCrime'] , cv=10, scoring='neg_mean_squared_error' ).mean())

 Mean MSE: -0.131116


ii.	What is the MSE on the training set (train on all the data then test on it all)?

In [254]:
mean_squared_error(df1['ViolentCrimesPerPop'], y_predict)

1.6958021953832663e-28

iii.	Does this mean the quadratic model is better than the linear model for this problem?

MSE for Polynomial Regression: -0.131116

MSE for Linear Regression: -0.020093969304445319

Mean Squared Error(MSE) for Polynomial and Linear Regressions are almost same. The results for polynomial regression are a bit better as more data points are covered by the polynomial curve.

# 4. Dirty Data

Read dirty dataset

Create a new field “highCrime” which is true if the crime rate per capita (ViolentCrimesPerPop) is greater than 0.1, and false otherwise.

In [255]:
d_data = pd.read_csv("E:\ML\Assignment1\Crime Prediction Data\communities-crime-full.csv")
d_data["highCrime"] = d_data["ViolentCrimesPerPop"] > 0.1

Create copy of dirty dataset

In [256]:
d_data1 = d_data.copy()

Remove unpredictive features from dataset

In [257]:
del d_data1['county']
del d_data1['community']
del d_data1['fold']
del d_data1['highCrime']
del d_data1['ViolentCrimesPerPop']
del d_data1['state']
del d_data1['communityname']

Replacing missing values with mean

In [258]:
for i in d_data1.columns:
    d_data1[[i]] = d_data1[[i]].convert_objects(convert_numeric = True)
    d_data1[[i]] = d_data1[[i]].fillna(d_data1[[i]].mean())

C:\Users\shrad\Anaconda3\lib\site-packages\ipykernel\__main__.py:2: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  from ipykernel import kernelapp as app


In [259]:
d_data1

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
5           0.02           0.28          0.06          0.54          1.00   
6           0.01           0.39          0.00          0.98          0.06   
7           0.01           0.74          0.03          0.46          0.20   
8           0.03           0.34          0.20          0.84          0.02   
9           0.01           0.40          0.06          0.87          0.30   
10          0.13           0.71          0.15          0.07          1.00   
11          0.02           0.46          0.08          0.91          0.07   
12          0.03           0.47          0.01          0.96          0.13   
13          0.01           0.44          0.00          0.98          0.04   
14          0.04           0.36          0.01          0.85          0.14   
15          0.03           0.34          0.06          0.93          0.03   
16          0.15           0.31          0.40          0.63          0.14   
17          0.01           0.53          0.01          0.94          0.20   
18          0.02           0.47          0.01          0.97          0.07   
19          0.00           0.41          0.05          0.96          0.01   
20          0.25           0.54          0.05          0.71          0.48   
21          1.00           0.42          0.47          0.59          0.12   
22          0.01           0.34          0.02          0.87          0.07   
23          0.11           0.43          0.04          0.89          0.09   
24          0.02           0.96          0.05          0.00          1.00   
25          0.00           0.33          0.02          0.91          0.16   
26          0.06           0.49          0.46          0.00          1.00   
27          0.01           0.37          0.01          0.99          0.02   
28          0.01           0.27          0.43          0.64          0.08   
29          0.00           0.50          0.01          0.98          0.02   
...          ...            ...           ...           ...           ...   
1964        0.09           0.45          1.00          0.08          0.11   
1965        1.00           0.36          0.44          0.64          0.06   
1966        0.09           0.41          0.07          0.74          0.39   
1967        0.01           0.66          0.00          0.97          0.10   
1968        0.00           0.24          0.51          0.53          0.11   
1969        1.00           0.29          0.21          0.29          1.00   
1970        0.00           0.37          0.38          0.69          0.04   
1971        0.06           0.39          0.32          0.73          0.04   
1972        0.08           0.49          0.07          0.69          0.27   
1973        0.00           0.38          0.03          0.97          0.04   
1974        0.00           0.41          0.62          0.51          0.02   
1975        0.05           0.27          0.03          0.89          0.27   
1976        0.06           0.29          0.03          0.93          0.18   
1977        0.01           0.35          0.06          0.94          0.06   
1978        0.01           0.54          0.02          0.94          0.13   
1979        0.02           0.43          0.39          0.63          0.15   
1980        0.06           0.43          0.04          0.93          0.03   
1981        0.07           0.38          0.17          0.84          0.11   
1982        0.00           0.47          0.02          0.98          0.03   
1983        0.01           0.45          0.37          0.65          0.15   
19

10 fold cross validation precision, recall and accuracy

In [260]:
d_model = DecisionTreeClassifier()
print("Precision: %f" % cross_val_score(d_model, d_data1, d_data['highCrime'], cv=10, scoring='precision').mean())
print ("Recall: %f" %cross_val_score(d_model, d_data1, d_data['highCrime'], cv=10, scoring='recall').mean())
print ("Accuracy: %f" %cross_val_score(d_model, d_data1, d_data['highCrime'], cv=10, scoring='accuracy').mean())

Precision: 0.809261
Recall: 0.788952
Accuracy: 0.754740


The training accuracy, precision and recall for dirty data

In [261]:
d_model.fit(d_data1, d_data['highCrime'])
d_predict = d_model.predict(d_data1)
print("Accuracy: %f" % d_model.score(d_data1, d_data['highCrime']))
print(precision_recall_fscore_support(d_data['highCrime'], d_predict))

Accuracy: 1.000000
(array([ 1.,  1.]), array([ 1.,  1.]), array([ 1.,  1.]), array([ 743, 1251], dtype=int64))


Top 10 features used for classification

In [262]:
total_col = d_data1.columns
d_imp_features = d_model.feature_importances_
index = np.argsort(d_imp_features)[::-1]
print('Feature Ranking')
for i in range(10):
        print("{}. Feature {} ({})".format(i+1, total_col[index[i]], d_imp_features[index[i]]))

Feature Ranking
1. Feature PctKids2Par (0.35879596764301586)
2. Feature racePctWhite (0.08962673869595336)
3. Feature racePctHisp (0.04857087695880409)
4. Feature PctEmplManu (0.022292672993011596)
5. Feature PctLess9thGrade (0.02066687029035681)
6. Feature PctEmploy (0.01665799433846146)
7. Feature blackPerCap (0.015445741602824008)
8. Feature PctHousOwnOcc (0.015290467997957283)
9. Feature MedRent (0.014716247970665456)
10. Feature PctWOFullPlumb (0.014144377961743941)


a.	Are the CV results better or worse? What does this say about the effect of missing values?

It is observed that the results with dirty data are better than the results for clean dataset. The missing values in the data are replaced by mean of the respective column as the Gaussian curve is normally distributed and gives best curve if mean is taken.
The possible reason for better results with dirty data is that the missing values could be more important which are now replaced with mean of the attribute.

# Extra Credit

In [263]:
#Random forest for clean data
model_rand = RandomForestClassifier(n_jobs=2)
model_rand.fit(df2, df1['highCrime'])

print("Precision: %f" % cross_val_score(model_rand , df2, df1['highCrime'], cv=10, scoring='precision' ).mean())
print("Recall: %f" % cross_val_score(model_rand , df2, df1['highCrime'], cv=10, scoring='recall' ).mean())
print("Accuracy: %f" % cross_val_score(model_rand , df2, df1['highCrime'], cv=10, scoring='accuracy' ).mean())


# Build a classification task using 3 informative features

imp_feature = model_rand.feature_importances_
indices = np.argsort(imp_feature)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{}. Feature {} ({})".format(i+1,col[indices[i]],imp_feature[indices[i]]))


Precision: 0.855610
Recall: 0.798400
Accuracy: 0.785178
Feature Ranking
1. Feature PctFam2Par (0.08452477089975259)
2. Feature PctPersDenseHous (0.062185659426804554)
3. Feature PctKids2Par (0.056136663790226984)
4. Feature NumIlleg (0.04749775117818757)
5. Feature FemalePctDiv (0.044990105808303336)
6. Feature racepctblack (0.03358638027312348)
7. Feature PctHousLess3BR (0.029028331099002946)
8. Feature PctPopUnderPov (0.028776635976397076)
9. Feature racePctWhite (0.027557833963627522)
10. Feature MalePctDivorce (0.02618377805051194)


In [264]:
#Random forest for dirty data
model_rand_dirty = RandomForestClassifier(n_jobs=2)
model_rand_dirty.fit(d_data1, d_data['highCrime'])

print ("Precision %f" %cross_val_score(model_rand_dirty, d_data1, d_data['highCrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(model_rand_dirty, d_data1, d_data['highCrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(model_rand_dirty, d_data1, d_data['highCrime'], cv=10, scoring='accuracy' ).mean())

#Important Features
imp_feature = model_rand_dirty.feature_importances_
indices = np.argsort(imp_feature)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{}. Feature {} ({})".format(i+1, total_col[indices[i]], imp_feature[indices[i]]))

Precision 0.872366
Recall 0.836940
Accuracy 0.809436
Feature Ranking
1. Feature PctIlleg (0.08182662872646537)
2. Feature PctKids2Par (0.06859735189185799)
3. Feature racePctWhite (0.05626670534717078)
4. Feature FemalePctDiv (0.052280499617489984)
5. Feature MalePctDivorce (0.040720538896956635)
6. Feature NumUnderPov (0.0346573898887012)
7. Feature TotalPctDiv (0.03173085931563321)
8. Feature NumIlleg (0.030458332410297495)
9. Feature PctPopUnderPov (0.029559841097819728)
10. Feature PctPersDenseHous (0.027456202715261453)


In [265]:
#perceptron for clean data
model_per = Perceptron()
model_per.fit(df2,df1['highCrime'])

print("Precision: %f" % cross_val_score(model_per, df2, df1['highCrime'],  cv=10, scoring='precision').mean())
print("Recall: %f" % cross_val_score(model_per, df2, df1['highCrime'], cv=10, scoring='recall').mean())
print("Accuracy: %f" % cross_val_score(model_per, df2, df1['highCrime'], cv=10,scoring='accuracy').mean())

#Important Features
imp_feature = model_per.coef_[0]
indices = np.argsort(imp_feature)[::-1]

print ('Feature Ranking') 
for i in range(10):
    print ("{}. Feature {} ({})".format(i+1, col[indices[i]], imp_feature[indices[i]]))

Precision: 0.879222
Recall: 0.779200
Accuracy: 0.787683
Feature Ranking
1. Feature racepctblack (16.71999999999997)
2. Feature racePctHisp (12.210000000000003)
3. Feature PctPersDenseHous (10.07000000000001)
4. Feature MalePctDivorce (9.340000000000018)
5. Feature PctIlleg (9.22)
6. Feature OtherPerCap (8.219999999999992)
7. Feature TotalPctDiv (7.679999999999988)
8. Feature FemalePctDiv (6.970000000000013)
9. Feature RentHighQ (6.960000000000011)
10. Feature HousVacant (6.779999999999991)


In [266]:
#perceptron for dirty data
model_per_dirty = Perceptron()
model_per_dirty.fit(d_data1, d_data['highCrime'])

print("Precision: %f" % cross_val_score(model_per_dirty, d_data1, d_data['highCrime'],  cv=10, scoring='precision').mean())
print("Recall: %f" % cross_val_score(model_per_dirty, d_data1, d_data['highCrime'], cv=10, scoring='recall').mean())
print("Accuracy: %f" % cross_val_score(model_per_dirty, d_data1, d_data['highCrime'], cv=10,scoring='accuracy').mean())


#Important Features
imp_feature = model_per_dirty.coef_[0]
indices = np.argsort(imp_feature)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{}. Feature {} ({})".format(i+1, total_col[indices[i]], imp_feature[indices[i]]))

Precision: 0.757628
Recall: 0.931200
Accuracy: 0.750300
Feature Ranking
1. Feature racepctblack (23.949999999999953)
2. Feature racePctHisp (10.400000000000006)
3. Feature OtherPerCap (9.559999999999999)
4. Feature PctPersDenseHous (9.210000000000022)
5. Feature HousVacant (8.880000000000019)
6. Feature PctIlleg (8.709999999999985)
7. Feature PctSameCity85 (8.330000000000009)
8. Feature RentHighQ (8.12000000000002)
9. Feature MalePctDivorce (7.869999999999992)
10. Feature PopDens (7.279999999999992)


i.	Experiment with two learning methods other than those described above (one can be a non-linear kernel for SVM) for the classification problem, explaining clearly what you did. Show CV results for both the clean and full datasets.

I used Random Forest and Perceptron methods for classification. I calcullated the precision, recall and accuracy using cross validation method for both Random Forest and Perceptron for clean as well as dirty data set.

ii.	What method gives the best results?

Accuarcy for clean data of Random Forest: 0.794229

Accuracy for clean data of Perceptron: 0.787683

Random Forest give better result than Perceptron on clean data set. Random Forest and Perceptron have almost same results.

Accuarcy for dirty data of Random Forest: 0.823441

Accuracy for dirty data of Perceptron: 0.737729

Random Forest give better result than Perceptron on clean data set.

Accuracy for clean data of Random Forest: 0.794229

Accuarcy for dirty data of Random Forest: 0.823441

Random Forest gives better result on dirty data because of reasons mentioned in above answer.

Accuracy for clean data of Perceptron: 0.787683

Accuracy for dirty data of Perceptron: 0.737729

Perceptron gives better result on clean data.